loadtest/loadtest.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package loadtest
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/mattermost/mattermost-load-test-ng/config"
+	"github.com/mattermost/mattermost-load-test-ng/loadtest/control"
+
+	"github.com/mattermost/mattermost-server/v5/mlog"
+)
+
+// LoadTester is a structure holding all the state needed to run a load-test
+type LoadTester struct {
+	controllers   []control.UserController
+	config        *config.LoadTestConfig
+	wg            sync.WaitGroup
+	status        chan control.UserStatus
+	started       bool
+	newController NewController
+}
+
+// NewController is a factory function that returns a new
+// control.UserController given an id and a channel of control.UserStatus
+type NewController func(int, chan<- control.UserStatus) control.UserController
+
+func (lt *LoadTester) handleStatus() {
+	for us := range lt.status {
+		if us.Code == control.USER_STATUS_STOPPED || us.Code == control.USER_STATUS_FAILED {
+			lt.wg.Done()
+		}
+		if us.Code == control.USER_STATUS_ERROR {
+			mlog.Info(us.Err.Error(), mlog.Int("controller_id", us.ControllerId))
+			continue
+		} else if us.Code == control.USER_STATUS_FAILED {
+			mlog.Error(us.Err.Error())
+			continue
+		}
+		mlog.Info(us.Info, mlog.Int("controller_id", us.ControllerId))
+	}
+}
+
+// AddUser increments by one the number of concurrently active users
+func (lt *LoadTester) AddUser() error {
+	if !lt.started {
+		return errors.New("LoadTester is not running")
```

Comment:
Let's use named errors for these. They can be reused in RemoveUser and Stop methods and also tested better in tests by avoiding a string match. Also, idiomatically speaking we should not capitalize error strings: https://github.com/golang/go/wiki/CodeReviewComments#error-strings. But I am fine with either ways, as long as we are consistent, which we are here.

Reply:
By *named errors* you mean defining constants like `ErrSomething = errors.New("some error")`?. If so we can definitely do that.  
Also, I normally use lowercase strings for errors but in this case it's the name of the type that's capitalized. Maybe we can rephrase if we really want to follow the guidelines.

Reply:
> By named errors you mean defining constants like ErrSomething = errors.New("some error")?. If so we can definitely do that.

Yep, I meant that.

> Also, I normally use lowercase strings for errors but in this case it's the name of the type that's capitalized

Ah ok, no worries :+1: